File: Week-Wise-Assignments/Assignment 10/Compressed Solutions/C Programs/2_graph_adjacency_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int vertex;
    struct Node* next;
} Node;

Node* adjList[100];
int vertices = 0;

Node* createNode(int v) {
    Node* n = malloc(sizeof(Node));
    n->vertex = v; n->next = NULL;
    return n;
}

void addEdge(int src, int dest) {
    if(src < vertices && dest < vertices) {
        Node* n1 = createNode(dest);
        n1->next = adjList[src];
        adjList[src] = n1;
        
        Node* n2 = createNode(src);
        n2->next = adjList[dest];
        adjList[dest] = n2;
    }
}

int countVertices() { return vertices; }

int countEdges() {
    int edges = 0;
    for(int i = 0; i < vertices; i++) {
        Node* temp = adjList[i];
        while(temp) { edges++; temp = temp->next; }
    }
    return edges / 2; // Undirected
}

void findAdjacent(int v) {
    printf("Adjacent to %d: ", v);
    Node* temp = adjList[v];
    while(temp) {
        printf("%d ", temp->vertex);
        temp = temp->next;
    }
    printf("\n");
}

int searchVertex(int v) {
    return v >= 0 && v < vertices;
}

void display() {
    for(int i = 0; i < vertices; i++) {
        printf("%d: ", i);
        Node* temp = adjList[i];
        while(temp) {
            printf("%d ", temp->vertex);
            temp = temp->next;
        }
        printf("\n");
    }
}

int main() {
    int ch, src, dest, v;
    while(1) {
        printf("\n1.Create 2.Add 3.Count V 4.Count E 5.Adjacent 6.Search 7.Display 0.Exit\n");
        scanf("%d", &ch);
        switch(ch) {
            case 1: printf("Vertices: "); scanf("%d", &vertices); 
                   for(int i = 0; i < vertices; i++) adjList[i] = NULL; break;
            case 2: printf("Edge: "); scanf("%d %d", &src, &dest); addEdge(src, dest); break;
            case 3: printf("Vertices: %d\n", countVertices()); break;
            case 4: printf("Edges: %d\n", countEdges()); break;
            case 5: printf("Vertex: "); scanf("%d", &v); findAdjacent(v); break;
            case 6: printf("Vertex: "); scanf("%d", &v); printf("%s\n", searchVertex(v) ? "Found" : "Not found"); break;
            case 7: display(); break;
            case 0: return 0;
        }
    }
}

File: Week-Wise-Assignments/Assignment 10/Compressed Solutions/Python Programs/2_graph_adjacency_list.py
from collections import defaultdict

class GraphList:
    def __init__(self):
        self.vertices = 0
        self.adj_list = defaultdict(list)
    
    def create(self, v):
        self.vertices = v
        self.adj_list = defaultdict(list)
    
    def add_edge(self, src, dest):
        if src < self.vertices and dest < self.vertices:
            self.adj_list[src].append(dest)
            self.adj_list[dest].append(src)
    
    def count_vertices(self): return self.vertices
    
    def count_edges(self):
        return sum(len(self.adj_list[i]) for i in range(self.vertices)) // 2
    
    def find_adjacent(self, v): return self.adj_list[v]
    
    def search_vertex(self, v): return 0 <= v < self.vertices
    
    def display(self):
        for i in range(self.vertices):
            print(f"{i}: {self.adj_list[i]}")

g = GraphList()
while True:
    print("\n1.Create 2.Add 3.Count V 4.Count E 5.Adjacent 6.Search 7.Display 0.Exit")
    ch = int(input())
    if ch == 1: g.create(int(input("Vertices: ")))
    elif ch == 2: src, dest = map(int, input("Edge: ").split()); g.add_edge(src, dest)
    elif ch == 3: print("Vertices:", g.count_vertices())
    elif ch == 4: print("Edges:", g.count_edges())
    elif ch == 5: print("Adjacent:", g.find_adjacent(int(input("Vertex: "))))
    elif ch == 6: print("Found" if g.search_vertex(int(input("Vertex: "))) else "Not found")
    elif ch == 7: g.display()
    elif ch == 0: break